File: examples/typescript/03-mcp-server-fastmcp-dpop/agent.ts
/**
 * Tier 03 — Agent with DPoP-bound tokens (TypeScript).
 *
 * Acquires a DPoP-bound (RFC 9449) access token from Authplane via the
 * `client_credentials` grant with both `mcp:echo` and `mcp:add` scopes, then
 * calls the tier-03 FastMCP server's `echo` and `add_numbers` tools. Each
 * outbound MCP request attaches a fresh DPoP proof bound to the same
 * keypair the AS bound the token to (`cnf.jkt`). The auth-specific lines
 * are wrapped between `// authplane:begin` / `// authplane:end` so
 * `tools/loccount` can audit the LOC budget for this tier.
 *
 * Run via the example's Makefile (the FastMCP server must already be up):
 *
 *     cp .env.example .env
 *     make run
 *     make verify
 */

// === transport / HTTP client boilerplate =====================================
import { generateKeyPairSync } from "node:crypto";
// MCP_URL is the network address the agent dials (Docker service DNS).
// DPOP_HTU is the URL the DPoP proof's `htu` claim must carry — and it is the
// resource origin (AUTHPLANE_RESOURCE), NOT the dial address: the FastMCP
// adapter reconstructs the expected `htu` from its own configured resource
// host (localhost:8080), so a proof signed for the Docker-DNS dial address
// (mcp-server:8080) would be rejected with "DPoP proof htu mismatch". RFC 9449 §4.2.
const MCP_URL = process.env.MCP_URL ?? "http://mcp-server:8080/mcp";
const DPOP_HTU = process.env.AUTHPLANE_RESOURCE ?? "http://localhost:8080/mcp";
const ECHO_TEXT = process.env.ECHO_TEXT ?? "hello from tier 03";

// Per-agent DPoP keypair. Plain Node crypto, no Authplane involvement.
const { privateKey } = generateKeyPairSync("ec", { namedCurve: "P-256" });
const pem = privateKey.export({ type: "pkcs8", format: "pem" }).toString();

// === Authplane integration ===================================================
// authplane:begin
import { AuthplaneClient, DPoPKeyMaterial, DPoPProvider } from "@authplane/sdk/core";
const dpop = new DPoPProvider({ keyMaterial: await DPoPKeyMaterial.fromPem(pem) });
// devMode relaxes the SDK's HTTPS/SSRF guard so the agent can discover metadata
// from the local http:// issuer. Local development only — production issuers must
// be https:// and devMode must be false (the default).
const ap = await AuthplaneClient.create({
  issuer: process.env.AUTHPLANE_ISSUER!,
  auth: { clientId: process.env.AUTHPLANE_CLIENT_ID!, clientSecret: process.env.AUTHPLANE_CLIENT_SECRET! },
  dpopProvider: dpop, devMode: true,
});
const token = await ap.clientCredentials(["mcp:echo", "mcp:add"], [process.env.AUTHPLANE_RESOURCE!]);
// authplane:end

// === call the MCP tools ======================================================
// MCP streamable-http session id. `initialize` returns it in the
// `Mcp-Session-Id` response header; every subsequent request (including the
// `notifications/initialized` handshake) must echo it back or the server
// rejects with 400 "No valid session ID provided".
let sessionId: string | undefined;

// `id` is omitted for JSON-RPC notifications (e.g. notifications/initialized),
// which carry no `id` and get a 202 with an empty body.
async function mcpCall(method: string, params: unknown, id?: number): Promise<unknown> {
  // authplane:begin
  const proof = await dpop.createProof({ method: "POST", url: DPOP_HTU, accessToken: token.accessToken });
  // Send the access token under the `Bearer` scheme (not RFC 9449's `DPoP`
  // scheme): FastMCP's bearer auth backend only inspects `Bearer `-prefixed
  // Authorization headers. The DPoP binding is still enforced — the server
  // verifies the access token's `cnf.jkt` against the `DPoP` proof header
  // regardless of the Authorization scheme.
  const authHeaders = { Authorization: `Bearer ${token.accessToken}`, DPoP: proof };
  // authplane:end
  const message: Record<string, unknown> = { jsonrpc: "2.0", method, params };
  if (id !== undefined) message.id = id;
  const res = await fetch(MCP_URL, {
    method: "POST",
    headers: {
      ...authHeaders,
      "Content-Type": "application/json",
      "Accept": "application/json, text/event-stream",
      ...(sessionId ? { "Mcp-Session-Id": sessionId } : {}),
    },
    body: JSON.stringify(message),
  });
  if (!res.ok) {
    throw new Error(`MCP ${method} returned HTTP ${res.status}: ${await res.text()}`);
  }
  // The server assigns the session on the initialize response; capture it for
  // every later request.
  const sid = res.headers.get("mcp-session-id");
  if (sid) sessionId = sid;
  // Notifications (no id) get 202 + empty body — nothing to parse.
  const body = await res.text();
  if (!body.trim()) return undefined;
  const dataLine = body.split("\n").find((l) => l.startsWith("data: "));
  const payload = dataLine ? dataLine.slice("data: ".length) : body;
  return JSON.parse(payload);
}

const initResult = await mcpCall("initialize", {
  protocolVersion: "2024-11-05",
  capabilities: {},
  clientInfo: { name: "tier-03-agent", version: "1.0" },
}, 1);
console.log(`[agent] initialize OK: ${JSON.stringify(initResult)}`);

// Complete the MCP lifecycle handshake: tools are not callable until the
// client confirms initialization. No id — this is a notification.
await mcpCall("notifications/initialized", {}, undefined);

const echoResult = await mcpCall("tools/call", {
  name: "echo",
  arguments: { text: ECHO_TEXT },
}, 2);
console.log(`[agent] echo OK: ${JSON.stringify(echoResult)}`);

const addResult = await mcpCall("tools/call", {
  name: "add_numbers",
  arguments: { a: 2, b: 3 },
}, 3);
console.log(`[agent] add_numbers OK: ${JSON.stringify(addResult)}`);

await ap.close();

File: examples/python/03-mcp-server-dpop-scopes/agent.py
"""Tier 03 — DPoP-bound agent (Python).

A minimal MCP client that acquires an Authplane-issued DPoP-bound machine
token (RFC 9449) via the `client_credentials` grant and uses it to call
two protected MCP tools (`echo` and `add_numbers`) on the tier-03 server.
Every outbound HTTP call to the AS and to the MCP server carries a fresh
`DPoP` proof header signed with an ephemeral EC key generated at startup.

The auth-specific code is wrapped between `# authplane:begin` /
`# authplane:end` so the `tools/loccount` tool can audit the LOC budget
for this tier.

Run via the example's Makefile (the tier-03 server must be up first):

    cp .env.example .env
    make run
    make verify
"""

# === transport / HTTP client boilerplate =====================================
import asyncio
import json
import os
import sys

import httpx
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import ec


async def call(http: httpx.AsyncClient, mcp_url: str, headers: dict[str, str],
               method: str, params: dict) -> tuple[httpx.Response, dict]:
    """JSON-RPC call helper. Returns (raw response, parsed dict).

    FastMCP's streamable-http transport can frame replies as raw JSON or as
    a single SSE `data:` event; this helper handles both shapes.
    """
    resp = await http.post(
        mcp_url, headers={**headers, "Content-Type": "application/json",
                          "Accept": "application/json, text/event-stream"},
        content=json.dumps({"jsonrpc": "2.0", "id": 1, "method": method, "params": params}),
    )
    if resp.status_code not in (200, 202):
        raise RuntimeError(f"MCP {method} failed: HTTP {resp.status_code}\n{resp.text}")
    body = resp.text
    # Strip a leading SSE `data:` prefix if present.
    for line in body.splitlines():
        if line.startswith("data: "):
            return resp, json.loads(line[len("data: "):])
    return resp, json.loads(body)


async def notify(http: httpx.AsyncClient, mcp_url: str, headers: dict[str, str],
                 method: str, params: dict | None = None) -> None:
    """JSON-RPC notification (no id, no response expected)."""
    payload: dict = {"jsonrpc": "2.0", "method": method}
    if params is not None:
        payload["params"] = params
    await http.post(
        mcp_url, headers={**headers, "Content-Type": "application/json",
                          "Accept": "application/json, text/event-stream"},
        content=json.dumps(payload),
    )


async def main() -> int:
    mcp_url = os.environ["MCP_URL"]

    # Generate an ephemeral EC P-256 keypair for DPoP proof signing. The key
    # never leaves this process; rotating it just means a fresh `cnf.jkt`
    # on the next token mint. PEM bytes are what `DPoPKeyMaterial.from_pem`
    # consumes (see python-sdk/authplane/dpop.py:106-116).
    priv_pem = ec.generate_private_key(ec.SECP256R1()).private_bytes(
        serialization.Encoding.PEM, serialization.PrivateFormat.PKCS8,
        serialization.NoEncryption(),
    )

    # === Authplane integration ===============================================
    # authplane:begin
    from authplane import ASCredentials, AuthplaneClient, DPoPKeyMaterial, DPoPProvider

    client = await AuthplaneClient.create(
        issuer=os.environ["AUTHPLANE_ISSUER"],
        auth=ASCredentials(os.environ["CLIENT_ID"], os.environ["CLIENT_SECRET"]),
        dpop=DPoPProvider(DPoPKeyMaterial.from_pem(priv_pem)), dev_mode=True,
    )
    token = (await client.client_credentials(
        scopes=["mcp:echo", "mcp:add"], resources=[os.environ["RESOURCE_URI"]],
    )).access_token
    # authplane:end

    # === call both MCP tools =================================================
    # `client.dpop_headers(...)` returns `{"DPoP": "<proof JWT>"}` signed with
    # the same key the access token is bound to and pinned to this exact
    # method+URL+access_token (see python-sdk/authplane/dpop.py:294-307).
    # RFC 9449 §7.1 specifies `Authorization: DPoP <token>`; we send the
    # token under the `Bearer` scheme because the upstream MCP SDK's
    # `BearerAuthBackend` only inspects headers starting with `bearer `
    # (mcp/server/auth/middleware/bearer_auth.py:35). The DPoP binding is
    # still enforced — the resource server verifies the `cnf.jkt` against
    # the `DPoP` proof header regardless of the access-token scheme.
    def auth_headers() -> dict[str, str]:
        h = client.dpop_headers("POST", mcp_url, access_token=token)
        h["Authorization"] = f"Bearer {token}"
        return h

    # An optional one-shot bearer-replay probe: if AGENT_DROP_DPOP_PROOF=1 is
    # set in the env, we strip the DPoP header before sending. The server's
    # verifier must reject this (HTTP 401) — that's the proof that DPoP is
    # actually being enforced and not just configured.
    drop = os.environ.get("AGENT_DROP_DPOP_PROOF") == "1"

    try:
        async with httpx.AsyncClient(timeout=10.0) as http:
            headers = auth_headers()
            if drop:
                headers.pop("DPoP", None)
            # 1) initialize — proves the bearer chain accepted the DPoP-bound
            # token AND the DPoP proof header bound to that token. The
            # streamable-http transport returns an `mcp-session-id` header on
            # this first request; all subsequent calls must include it.
            init_resp, init = await call(http, mcp_url, headers, "initialize", {
                "protocolVersion": "2024-11-05", "capabilities": {},
                "clientInfo": {"name": "tier03-agent", "version": "1.0"},
            })
            print(json.dumps(init, separators=(",", ":")))
            if "demo-server" not in json.dumps(init):
                print("initialize response missing serverInfo.name", file=sys.stderr)
                return 1
            session_id = init_resp.headers.get("mcp-session-id")
            session_h = {"mcp-session-id": session_id} if session_id else {}

            # Per MCP spec the client must follow `initialize` with a
            # `notifications/initialized` notification before issuing tool
            # calls.
            await notify(http, mcp_url, {**auth_headers(), **session_h},
                         "notifications/initialized")

            # 2) tools/call: echo  — requires the mcp:echo scope
            _, echo_resp = await call(http, mcp_url, {**auth_headers(), **session_h},
                                      "tools/call",
                                      {"name": "echo", "arguments": {"text": "hello"}})
            print(json.dumps(echo_resp, separators=(",", ":")))
            if "hello" not in json.dumps(echo_resp):
                print("echo response did not echo 'hello'", file=sys.stderr)
                return 1

            # 3) tools/call: add_numbers  — requires the mcp:add scope
            _, add_resp = await call(http, mcp_url, {**auth_headers(), **session_h},
                                     "tools/call",
                                     {"name": "add_numbers", "arguments": {"a": 2, "b": 3}})
            print(json.dumps(add_resp, separators=(",", ":")))
            if '"5"' not in json.dumps(add_resp) and ":5" not in json.dumps(add_resp):
                print("add_numbers response did not contain 5", file=sys.stderr)
                return 1

        print("authenticated MCP DPoP + per-tool scope calls OK", flush=True)
        return 0
    finally:
        await client.aclose()


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))
